# Laboratorio 1

# Preparazione del dato: pulizia e integrazione (prova aggiornamento da Binder)

06/03/2024

In questo laboratorio impareremo a trattare dati provenienti da sorgenti diverse, cercando di ottenere informazioni integrate e complete. 
Come esempio, ci concentreremo su dati riguardanti film e serie tv dei maggiori servizi di streaming. 

In [1]:
import pandas as pd

### Carico i dati di Netflix

In [2]:
netflix_titles = pd.read_csv("netflix_titles.csv")

Diamo un'occhiata a come è organizzato il dataset

In [3]:
netflix_titles.head()

Unnamed: 0        id                                title   type  \
0           0  ts300399  Five Came Back: The Reference Films   SHOW   
1           1   tm82169                                Rocky  MOVIE   
2           2   tm17823                               Grease  MOVIE   
3           3   tm69975                             Rocky II  MOVIE   
4           4  tm127384      Monty Python and the Holy Grail  MOVIE   

   release_year age_certification  runtime                 genres  \
0          1945             TV-MA       51      ['documentation']   
1          1976                PG      119     ['drama', 'sport']   
2          1978                PG      110  ['comedy', 'romance']   
3          1979                PG      119     ['drama', 'sport']   
4          1975                PG       91  ['fantasy', 'comedy']   

  production_countries  seasons  ... tmdb_popularity  tmdb_score  \
0               ['US']      1.0  ...           0.601         NaN   
1               ['US']      NaN  ...          99.713       7.783   
2               ['US']      NaN  ...          33.160       7.406   
3               ['US']      NaN  ...          74.191       7.200   
4               ['GB']      NaN  ...          20.964       7.804   

                     director  \
0                         NaN   
1            John G. Avildsen   
2              Randal Kleiser   
3          Sylvester Stallone   
4  Terry Gilliam, Terry Jones   

                                                cast         country  \
0                                                NaN   United States   
1  Sylvester Stallone, Talia Shire, Burt Young, C...   United States   
2  John Travolta, Olivia Newton-John, Stockard Ch...   United States   
3  Sylvester Stallone, Talia Shire, Burt Young, C...   United States   
4  Graham Chapman, John Cleese, Eric Idle, Terry ...  United Kingdom   

         date_added rating  duration  \
0  February 9, 2021  TV-MA  1 Season   
1    August 1, 2019     PG   120 min   
2  November 1, 2019     PG   110 min   
3    August 1, 2019     PG   119 min   
4   January 2, 2019     PG    92 min   

                                      listed_in  \
0                                      TV Shows   
1    Action & Adventure, Classic Movies, Dramas   
2    Classic Movies, Comedies, Music & Musicals   
3                         Dramas, Sports Movies   
4  Action & Adventure, Classic Movies, Comedies   

                                         description  
0  This collection includes 12 World War II-era p...  
1  Sylvester Stallone shot to fame as Rocky Balbo...  
2  John Travolta and Olivia Newton-John star in t...  
3  Featuring a rousing climax, this engaging sequ...  
4  The Monty Python comedy clan skewers King Arth...  

[5 rows x 23 columns]

ed allo schema (notate che con head() alcune delle colonne, per problemi di spazio, non compaiono)

In [4]:
netflix_titles.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3645 entries, 0 to 3644
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Unnamed: 0            3645 non-null   int64  
 1   id                    3645 non-null   object 
 2   title                 3645 non-null   object 
 3   type                  3645 non-null   object 
 4   release_year          3645 non-null   int64  
 5   age_certification     2003 non-null   object 
 6   runtime               3645 non-null   int64  
 7   genres                3645 non-null   object 
 8   production_countries  3645 non-null   object 
 9   seasons               1453 non-null   float64
 10  imdb_id               3417 non-null   object 
 11  imdb_score            3378 non-null   float64
 12  imdb_votes            3374 non-null   float64
 13  tmdb_popularity       3612 non-null   float64
 14  tmdb_score            3529 non-null   float64
 15  director             

### Carico i dati di Disney+

In [5]:
disney_titles = pd.read_csv("disney_titles.csv")

In [6]:
disney_titles.head()

Unnamed: 0        id                                   title show_type  \
0           0   tm61729  The Adventures of Ichabod and Mr. Toad     MOVIE   
1           1   tm61052                              Cinderella     MOVIE   
2           2  tm190870                              Cinderella     MOVIE   
3           3   tm87946                                   Dumbo     MOVIE   
4           4  tm207119                                   Dumbo     MOVIE   

   release_year age_certification  runtime  \
0          1950                 G       68   
1          1950                 G       74   
2          1997                 G       88   
3          1941                 G       64   
4          2019                PG      112   

                                              genres production_countries  \
0  ['horror', 'fantasy', 'animation', 'comedy', '...               ['US']   
1      ['fantasy', 'animation', 'family', 'romance']               ['US']   
2                   ['romance', 'family', 'fantasy']               ['US']   
3        ['animation', 'drama', 'family', 'fantasy']               ['US']   
4                    ['family', 'fantasy', 'action']         ['GB', 'US']   

   seasons    imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  \
0      NaN  tt0041094         6.9     16497.0           20.116       6.500   
1      NaN  tt0042332         7.3    164222.0           93.547       7.035   
2      NaN  tt0128996         6.7      8944.0            9.493       6.400   
3      NaN  tt0033563         7.2    134997.0           54.265       6.999   
4      NaN  tt3861390         6.3     73010.0           23.271       6.661   

                                          director  \
0         Jack Kinney, Clyde Geronimi, James Algar   
1  Wilfred Jackson, Hamilton Luske, Clyde Geronimi   
2  Wilfred Jackson, Hamilton Luske, Clyde Geronimi   
3                                       Tim Burton   
4                                       Tim Burton   

                                            country               from rating  \
0                                     United States  November 12, 2019      G   
1                                     United States  November 12, 2019      G   
2                                     United States  November 12, 2019      G   
3  United States, United Kingdom, Australia, Canada  November 12, 2019     PG   
4  United States, United Kingdom, Australia, Canada  November 12, 2019     PG   

  duration  
0   69 min  
1   77 min  
2   77 min  
3  117 min  
4  117 min

In [7]:
disney_titles.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1070 entries, 0 to 1069
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Unnamed: 0            1070 non-null   int64  
 1   id                    1070 non-null   object 
 2   title                 1070 non-null   object 
 3   show_type             1070 non-null   object 
 4   release_year          1070 non-null   int64  
 5   age_certification     854 non-null    object 
 6   runtime               1070 non-null   int64  
 7   genres                1070 non-null   object 
 8   production_countries  1070 non-null   object 
 9   seasons               263 non-null    float64
 10  imdb_id               807 non-null    object 
 11  imdb_score            792 non-null    float64
 12  imdb_votes            791 non-null    float64
 13  tmdb_popularity       1067 non-null   float64
 14  tmdb_score            1027 non-null   float64
 15  director             

Come abbiamo visto, i due dataset hanno molte colonne in comune, ma anche alcune differenze.

Proviamo comunque ad integrare le due tabelle, creandone una terza che metta insieme i dati della prima e quelli della seconda. 

**tip**
Dovete "concatenare" i due dataFrame di pandas. 

Provate a cercare la funzione corretta e scrivete il codice necessario. 

In [8]:
integrated = pd.concat([disney_titles, netflix_titles])
integrated.head()

Unnamed: 0        id                                   title show_type  \
0           0   tm61729  The Adventures of Ichabod and Mr. Toad     MOVIE   
1           1   tm61052                              Cinderella     MOVIE   
2           2  tm190870                              Cinderella     MOVIE   
3           3   tm87946                                   Dumbo     MOVIE   
4           4  tm207119                                   Dumbo     MOVIE   

   release_year age_certification  runtime  \
0          1950                 G       68   
1          1950                 G       74   
2          1997                 G       88   
3          1941                 G       64   
4          2019                PG      112   

                                              genres production_countries  \
0  ['horror', 'fantasy', 'animation', 'comedy', '...               ['US']   
1      ['fantasy', 'animation', 'family', 'romance']               ['US']   
2                   ['romance', 'family', 'fantasy']               ['US']   
3        ['animation', 'drama', 'family', 'fantasy']               ['US']   
4                    ['family', 'fantasy', 'action']         ['GB', 'US']   

   seasons  ...                                         director  \
0      NaN  ...         Jack Kinney, Clyde Geronimi, James Algar   
1      NaN  ...  Wilfred Jackson, Hamilton Luske, Clyde Geronimi   
2      NaN  ...  Wilfred Jackson, Hamilton Luske, Clyde Geronimi   
3      NaN  ...                                       Tim Burton   
4      NaN  ...                                       Tim Burton   

                                            country               from  \
0                                     United States  November 12, 2019   
1                                     United States  November 12, 2019   
2                                     United States  November 12, 2019   
3  United States, United Kingdom, Australia, Canada  November 12, 2019   
4  United States, United Kingdom, Australia, Canada  November 12, 2019   

   rating  duration type cast date_added listed_in description  
0       G    69 min  NaN  NaN        NaN       NaN         NaN  
1       G    77 min  NaN  NaN        NaN       NaN         NaN  
2       G    77 min  NaN  NaN        NaN       NaN         NaN  
3      PG   117 min  NaN  NaN        NaN       NaN         NaN  
4      PG   117 min  NaN  NaN        NaN       NaN         NaN  

[5 rows x 25 columns]

Adesso, stampiamo un po' di informazioni sul dataset integrato (come ad esempio le dimensioni) e diamo un'occhiata allo schema.

In [9]:
print("Dimensioni del dataset integrato:", integrated.shape)
print("Schema del dataset integrato:")
integrated.info()

Dimensioni del dataset integrato: (4715, 25)
Schema del dataset integrato:
<class 'pandas.core.frame.DataFrame'>
Index: 4715 entries, 0 to 3644
Data columns (total 25 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Unnamed: 0            4715 non-null   int64  
 1   id                    4715 non-null   object 
 2   title                 4715 non-null   object 
 3   show_type             1070 non-null   object 
 4   release_year          4715 non-null   int64  
 5   age_certification     2857 non-null   object 
 6   runtime               4715 non-null   int64  
 7   genres                4715 non-null   object 
 8   production_countries  4715 non-null   object 
 9   seasons               1716 non-null   float64
 10  imdb_id               4224 non-null   object 
 11  imdb_score            4170 non-null   float64
 12  imdb_votes            4165 non-null   float64
 13  tmdb_popularity       4679 non-null   float64
 14  tm

Nonostante le differenze tra le due tabelle, la funzione concat concatena le due tabelle, mettendo a Null i valori delle colonne presenti solo in una delle due ed in corrispondenza delle righe provenienti dal dataset in cui la colonna non è presente. Proviamo a verificare la presenza di Null

In [10]:
null_values = integrated.isnull().sum()
print("Valori Null nel dataset integrato", null_values)

Valori Null nel dataset integrato Unnamed: 0                 0
id                         0
title                      0
show_type               3645
release_year               0
age_certification       1858
runtime                    0
genres                     0
production_countries       0
seasons                 2999
imdb_id                  491
imdb_score               545
imdb_votes               550
tmdb_popularity           36
tmdb_score               159
director                1695
country                  505
from                    3646
rating                     2
duration                   1
type                    1070
cast                    1435
date_added              1070
listed_in               1070
description             1070
dtype: int64


Ad esempio la colonna **listed_in**, che è presente solo in netflix_titles, ha 1070 valori Null, che corrisponde al numero di righe della tabella disney_titles. 

Decidiamo di eliminare eventuali colonne non comuni. Possiamo farlo **prima** dell'integrazione, oppure direttamente sulla tabella concatenata. 

**Tip**: confrontate i due schemi per capire quali siano le colonne non comuni. Attenzione!!! Potrebbero anche esserci colonne che sembrano non comuni, ma in realtà contengono la stessa tipologia di informazione e hanno solo nomi diversi!!

In [11]:
disney_columns = set(disney_titles.columns)
netflix_columns = set(netflix_titles.columns)
# Troviamo le colonne non comuni (almeno a livello di nome)
non_common_columns = disney_columns.symmetric_difference(netflix_columns)
print("Colonne non comuni:", non_common_columns)
print("Dataset a confronto:")
disney_titles.info()
print("\n")
netflix_titles.info()

Colonne non comuni: {'type', 'date_added', 'listed_in', 'from', 'cast', 'show_type', 'description'}
Dataset a confronto:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1070 entries, 0 to 1069
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Unnamed: 0            1070 non-null   int64  
 1   id                    1070 non-null   object 
 2   title                 1070 non-null   object 
 3   show_type             1070 non-null   object 
 4   release_year          1070 non-null   int64  
 5   age_certification     854 non-null    object 
 6   runtime               1070 non-null   int64  
 7   genres                1070 non-null   object 
 8   production_countries  1070 non-null   object 
 9   seasons               263 non-null    float64
 10  imdb_id               807 non-null    object 
 11  imdb_score            792 non-null    float64
 12  imdb_votes            791 non-null    float64
 13  tm

Andiamo ora alla ricerca di eventuali colonne che contengono la stessa informazione. Guardate in particolare (type, show_type),  (contry, production_countries), (date_added, from)

In [12]:
# andiamo a rinominare dentro a disney_titles la colonna show_type in type e la colonna from in date_added
# così da unirle nel dataset integrato:
disney_titles.rename(columns={'show_type': 'type'}, inplace=True)
disney_titles.rename(columns={'from': 'date_added'}, inplace=True)

In [13]:
# Concatena i due DataFrame
integrated = pd.concat([disney_titles, netflix_titles])

# Verifichiamo che la colonna 'type' e la colonna 'date_added' sia stata unificata
print(integrated.columns)
integrated.info()

Index(['Unnamed: 0', 'id', 'title', 'type', 'release_year',
       'age_certification', 'runtime', 'genres', 'production_countries',
       'seasons', 'imdb_id', 'imdb_score', 'imdb_votes', 'tmdb_popularity',
       'tmdb_score', 'director', 'country', 'date_added', 'rating', 'duration',
       'cast', 'listed_in', 'description'],
      dtype='object')
<class 'pandas.core.frame.DataFrame'>
Index: 4715 entries, 0 to 3644
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Unnamed: 0            4715 non-null   int64  
 1   id                    4715 non-null   object 
 2   title                 4715 non-null   object 
 3   type                  4715 non-null   object 
 4   release_year          4715 non-null   int64  
 5   age_certification     2857 non-null   object 
 6   runtime               4715 non-null   int64  
 7   genres                4715 non-null   object 
 8   production_countries  4715 n

Infine, ci sono colonne che già ora possiamo giudicare inutili?

per buttare via una colonna, usiamo la funzione drop di pandas 

In [14]:
# immagino che Unnamed: 0 sia abbastanza inutile ai fini di una analisi statistica
# Rimuoviamo la colonna 'Unnamed: 0' dal DataFrame integrato
integrated.drop(columns=['Unnamed: 0'], inplace=True)

# inoltre, analizzando i dati, notiamo che le colonne age_certification e rating (presenti in entrambi i dataset, e quindi
# anche in quello integrato) contengono più o meno la stessa informazione ma in entrambi i casi la colonna rating è
# più completa, in quanto contiene la quasi totalità dei valori non nulli, mentre age_certification contiene molti nulli
# droppiamo quindi la colonna age_certification
# discorso simile per country e production_countries: stessa informazione (il/i Paese/i in cui è stato prodotto il contenuto )
# in production_countries l'informazione viene veicolata attraverso un array di sigle dei Paesi in questione,
# mentre in country con una lista dei nomi dei per esteso dei Paesi in questione
# analizzando il dataset vediamo che la colonna production_countries è più completa, non contenendo praticamente valori nulli
# quindi droppiamo country
# visualizziamo "quello che resta"
integrated.info()


<class 'pandas.core.frame.DataFrame'>
Index: 4715 entries, 0 to 3644
Data columns (total 22 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    4715 non-null   object 
 1   title                 4715 non-null   object 
 2   type                  4715 non-null   object 
 3   release_year          4715 non-null   int64  
 4   age_certification     2857 non-null   object 
 5   runtime               4715 non-null   int64  
 6   genres                4715 non-null   object 
 7   production_countries  4715 non-null   object 
 8   seasons               1716 non-null   float64
 9   imdb_id               4224 non-null   object 
 10  imdb_score            4170 non-null   float64
 11  imdb_votes            4165 non-null   float64
 12  tmdb_popularity       4679 non-null   float64
 13  tmdb_score            4556 non-null   float64
 14  director              3020 non-null   object 
 15  country               4210

### Carichiamo altre due tabelle, che forniscono fonti di informazioni aggiuntive

In [15]:
netflix_credits = pd.read_csv("netflix_credits.csv")
disney_credits = pd.read_csv("disney_credits.csv")

come sempre, stampiamo le prime righe di ciascun dataFrame 

In [16]:
netflix_credits.head()


person_id       id                name                  character   role
0       3180  tm82169  Sylvester Stallone      Robert "Rocky" Balboa  ACTOR
1       7748  tm82169         Talia Shire  Adrianna "Adrian" Pennino  ACTOR
2       7749  tm82169          Burt Young             Paulie Pennino  ACTOR
3       7750  tm82169       Carl Weathers               Apollo Creed  ACTOR
4       7751  tm82169    Burgess Meredith            Mickey Goldmill  ACTOR

In [17]:
disney_credits.head()

person_id       id            name         character   role
0      35549  tm89464  Maureen O'Hara      Doris Walker  ACTOR
1      57832  tm89464      John Payne       Fred Gailey  ACTOR
2      57833  tm89464    Edmund Gwenn      Kris Kringle  ACTOR
3      25096  tm89464    Natalie Wood      Susan Walker  ACTOR
4      27185  tm89464     Porter Hall  Granville Sawyer  ACTOR

Possiamo osservare che le tabelle credits e titles, hanno informazioni complementari.  In particolare, credits riporta delle informazinoi aggiuntive riguardanti attori e carattere interpretato. 

Come prima, procediamo con l'integrazione delle due tabelle, che questa volta è più facile (lo schema è proprio lo stesso!)

In [18]:
integrated_credits = pd.concat([disney_credits, netflix_credits])
integrated_credits.info()

<class 'pandas.core.frame.DataFrame'>
Index: 110791 entries, 0 to 80382
Data columns (total 5 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   person_id  110791 non-null  int64 
 1   id         110791 non-null  object
 2   name       110791 non-null  object
 3   character  98256 non-null   object
 4   role       110791 non-null  object
dtypes: int64(1), object(4)
memory usage: 5.1+ MB


Data la complementarietà delle informazioni che troviamo nelle cartelle titles e credits, possiamo decidere di fare un **join** tra le due tabelle, per includere tutte le informazioni in un unico dataset. 


In [22]:
# ...completare qui usando la funzione merge per fare il join
# integrated è la tabella che integra netflix_titles e disney_titles su cui abbiamo lavorato prima
# mentre integrated_credits è la tabella che integra disney_credits e netflix_credits
# le uniamo con la merge
final_table= pd.merge(integrated, integrated_credits, on='id', how='inner')

verifichiamo cosa sia successo guardando le dimensioni di ciascun dataframe

In [26]:
#completare qui
print("dimensioni della tabella 'totale'",final_table.shape)
print("dimensioni della tabella integrata dei credits", integrated_credits.shape)
print("dimensioni della tabella integrata dei titoli", integrated.shape)

dimensioni della tabella 'totale' (66189, 26)
dimensioni della tabella integrata dei credits (110791, 5)
dimensioni della tabella integrata dei titoli (4715, 22)


vi tornano le dimensioni?

In [27]:
final_table.head()

id                                   title   type  release_year  \
0  tm61729  The Adventures of Ichabod and Mr. Toad  MOVIE          1950   
1  tm61729  The Adventures of Ichabod and Mr. Toad  MOVIE          1950   
2  tm61729  The Adventures of Ichabod and Mr. Toad  MOVIE          1950   
3  tm61729  The Adventures of Ichabod and Mr. Toad  MOVIE          1950   
4  tm61729  The Adventures of Ichabod and Mr. Toad  MOVIE          1950   

  age_certification  runtime  \
0                 G       68   
1                 G       68   
2                 G       68   
3                 G       68   
4                 G       68   

                                              genres production_countries  \
0  ['horror', 'fantasy', 'animation', 'comedy', '...               ['US']   
1  ['horror', 'fantasy', 'animation', 'comedy', '...               ['US']   
2  ['horror', 'fantasy', 'animation', 'comedy', '...               ['US']   
3  ['horror', 'fantasy', 'animation', 'comedy', '...               ['US']   
4  ['horror', 'fantasy', 'animation', 'comedy', '...               ['US']   

   seasons    imdb_id  ...         date_added  rating  duration  cast  \
0      NaN  tt0041094  ...  November 12, 2019       G    69 min   NaN   
1      NaN  tt0041094  ...  November 12, 2019       G    69 min   NaN   
2      NaN  tt0041094  ...  November 12, 2019       G    69 min   NaN   
3      NaN  tt0041094  ...  November 12, 2019       G    69 min   NaN   
4      NaN  tt0041094  ...  November 12, 2019       G    69 min   NaN   

  listed_in description person_id             name  \
0       NaN         NaN     25056      Bing Crosby   
1       NaN         NaN     25058   Basil Rathbone   
2       NaN         NaN     25057       Eric Blore   
3       NaN         NaN      5901  J. Pat O'Malley   
4       NaN         NaN    598861     John McLeish   

                                           character   role  
0   Narrator (segment "The Legend of Sleepy Hollow")  ACTOR  
1  Narrator (segment "The Wind in the Willows") (...  ACTOR  
2                                   Mr. Toad (voice)  ACTOR  
3                          Cyril Proudbottom (voice)  ACTOR  
4                                 Prosecutor (voice)  ACTOR  

[5 rows x 26 columns]

In [28]:
print(final_table.columns)

Index(['id', 'title', 'type', 'release_year', 'age_certification', 'runtime',
       'genres', 'production_countries', 'seasons', 'imdb_id', 'imdb_score',
       'imdb_votes', 'tmdb_popularity', 'tmdb_score', 'director', 'country',
       'date_added', 'rating', 'duration', 'cast', 'listed_in', 'description',
       'person_id', 'name', 'character', 'role'],
      dtype='object')


e visualizziamo gli eventuali elementi mancanti

In [29]:
final_table.isnull().sum()

id                          0
title                       0
type                        0
release_year                0
age_certification       22496
runtime                     0
genres                      0
production_countries        0
seasons                 54660
imdb_id                  3968
imdb_score               4251
imdb_votes               4288
tmdb_popularity             0
tmdb_score                525
director                11698
country                  4063
date_added                  1
rating                     30
duration                    2
cast                    21400
listed_in               19352
description             19352
person_id                   0
name                        0
character                7059
role                        0
dtype: int64

concentriamoci sulla colonna date_added, intanto visualizzandola:

In [30]:
print(final_table['date_added'])

0        November 12, 2019
1        November 12, 2019
2        November 12, 2019
3        November 12, 2019
4        November 12, 2019
               ...        
66184        July 22, 2021
66185        July 22, 2021
66186        July 22, 2021
66187        July 22, 2021
66188        July 22, 2021
Name: date_added, Length: 66189, dtype: object


vediamo che giorni mesi e anni sono presenti. Potrebbe essere utile, però, separare queste informazioni, incorporandole in tre colonne diverse. 

Facciamolo insieme, con due diversi passi:

1. dobbiamo trasformare questa colonna in formato 'datetime' utilizzando la funzione di pandas 'to_datetime'

In [31]:
final_table['date_added'] = pd.to_datetime(final_table['date_added'])


In [32]:
print(final_table['date_added'])

0       2019-11-12
1       2019-11-12
2       2019-11-12
3       2019-11-12
4       2019-11-12
           ...    
66184   2021-07-22
66185   2021-07-22
66186   2021-07-22
66187   2021-07-22
66188   2021-07-22
Name: date_added, Length: 66189, dtype: datetime64[ns]


poi, possiamo separare le varie entrate in tre nuove colonne, che si chiameranno day, month e year. Bisogna fare colonna.dt.year e cambiare poi l'ultimo metodo 

In [43]:
final_table['day'] = final_table['date_added'].dt.day.fillna(0).astype(int)
final_table['month'] = final_table['date_added'].dt.month.fillna(0).astype(int)
final_table['year'] = final_table['date_added'].dt.year.fillna(0).astype(int)


ora visualizziamo il risultato

In [44]:
final_table.head()

id                                   title   type  release_year  \
0  tm61729  The Adventures of Ichabod and Mr. Toad  MOVIE          1950   
1  tm61729  The Adventures of Ichabod and Mr. Toad  MOVIE          1950   
2  tm61729  The Adventures of Ichabod and Mr. Toad  MOVIE          1950   
3  tm61729  The Adventures of Ichabod and Mr. Toad  MOVIE          1950   
4  tm61729  The Adventures of Ichabod and Mr. Toad  MOVIE          1950   

  age_certification  runtime  \
0                 G       68   
1                 G       68   
2                 G       68   
3                 G       68   
4                 G       68   

                                              genres production_countries  \
0  ['horror', 'fantasy', 'animation', 'comedy', '...               ['US']   
1  ['horror', 'fantasy', 'animation', 'comedy', '...               ['US']   
2  ['horror', 'fantasy', 'animation', 'comedy', '...               ['US']   
3  ['horror', 'fantasy', 'animation', 'comedy', '...               ['US']   
4  ['horror', 'fantasy', 'animation', 'comedy', '...               ['US']   

   seasons    imdb_id  ...  cast  listed_in  description  person_id  \
0      NaN  tt0041094  ...   NaN        NaN          NaN      25056   
1      NaN  tt0041094  ...   NaN        NaN          NaN      25058   
2      NaN  tt0041094  ...   NaN        NaN          NaN      25057   
3      NaN  tt0041094  ...   NaN        NaN          NaN       5901   
4      NaN  tt0041094  ...   NaN        NaN          NaN     598861   

              name                                          character   role  \
0      Bing Crosby   Narrator (segment "The Legend of Sleepy Hollow")  ACTOR   
1   Basil Rathbone  Narrator (segment "The Wind in the Willows") (...  ACTOR   
2       Eric Blore                                   Mr. Toad (voice)  ACTOR   
3  J. Pat O'Malley                          Cyril Proudbottom (voice)  ACTOR   
4     John McLeish                                 Prosecutor (voice)  ACTOR   

  day month  year  
0  12    11  2019  
1  12    11  2019  
2  12    11  2019  
3  12    11  2019  
4  12    11  2019  

[5 rows x 29 columns]

ora creiamo una copia della tabella, contenente soltanto i film.


In [48]:
movies_table = final_table[final_table['type'] == 'MOVIE'].copy()
print("solo film: ", movies_table.shape)
print("anche altri tipi di show: ", final_table.shape)
print("visualizziamo le prime 5 righe della nuova tabella con i film")
movies_table.head()

solo film:  (54660, 29)
anche altri tipi di show:  (66189, 29)
visualizziamo le prime 5 righe della nuova tabella con i film


id                                   title   type  release_year  \
0  tm61729  The Adventures of Ichabod and Mr. Toad  MOVIE          1950   
1  tm61729  The Adventures of Ichabod and Mr. Toad  MOVIE          1950   
2  tm61729  The Adventures of Ichabod and Mr. Toad  MOVIE          1950   
3  tm61729  The Adventures of Ichabod and Mr. Toad  MOVIE          1950   
4  tm61729  The Adventures of Ichabod and Mr. Toad  MOVIE          1950   

  age_certification  runtime  \
0                 G       68   
1                 G       68   
2                 G       68   
3                 G       68   
4                 G       68   

                                              genres production_countries  \
0  ['horror', 'fantasy', 'animation', 'comedy', '...               ['US']   
1  ['horror', 'fantasy', 'animation', 'comedy', '...               ['US']   
2  ['horror', 'fantasy', 'animation', 'comedy', '...               ['US']   
3  ['horror', 'fantasy', 'animation', 'comedy', '...               ['US']   
4  ['horror', 'fantasy', 'animation', 'comedy', '...               ['US']   

   seasons    imdb_id  ...  cast  listed_in  description  person_id  \
0      NaN  tt0041094  ...   NaN        NaN          NaN      25056   
1      NaN  tt0041094  ...   NaN        NaN          NaN      25058   
2      NaN  tt0041094  ...   NaN        NaN          NaN      25057   
3      NaN  tt0041094  ...   NaN        NaN          NaN       5901   
4      NaN  tt0041094  ...   NaN        NaN          NaN     598861   

              name                                          character   role  \
0      Bing Crosby   Narrator (segment "The Legend of Sleepy Hollow")  ACTOR   
1   Basil Rathbone  Narrator (segment "The Wind in the Willows") (...  ACTOR   
2       Eric Blore                                   Mr. Toad (voice)  ACTOR   
3  J. Pat O'Malley                          Cyril Proudbottom (voice)  ACTOR   
4     John McLeish                                 Prosecutor (voice)  ACTOR   

  day month  year  
0  12    11  2019  
1  12    11  2019  
2  12    11  2019  
3  12    11  2019  
4  12    11  2019  

[5 rows x 29 columns]

a questo punto, possiamo modificare la durata in ore, per prima cosa eliminiamo la parola minuti

In [49]:
movies_table['duration'] = movies_table['duration'].str.replace('min', '').astype(float)


ValueError: could not convert string to float: '1 Season'

quindi dividere per 60 (dopo conversione float)

In [ ]:
movies_table['duration'] /= 60

e finalmente, aggiungere un suffisso per indicare le ore, per esempio 'hours', tornando a stringa

In [ ]:
movies_table['duration'] = movies_table['duration'].astype(str) + ' hours'

stampiamo il risultato finale 

In [ ]:
print(movies['duration'])

concentriamoci sulla colonna release_year. Innanzitutto eliminiamo le righe dove questa informazinoe non sia presente. Usiamo la funzione dropna di pandas 
dropna(axis = , subset = [nome colonna])

In [ ]:
#completare qui 

In [ ]:
final_table_clean.isnull().sum()